279 add multiple argument to filter_var ChoicesFilterState and LogicalFilterState

kartikeyakirar opened the pull request:
In ChoicesFilterState's initialize and LogicalFilterState's initialize methods, _**multiple**_ arguments have been added. These arguments should be included in the return statement of the get_state method.

Furthermore, if the user selects "multiple = F", the checkboxgroupinput should be replaced with a radiobutton.

closes #250 #279

m7pr commented:
@kartikeyakirar shouldn't this be merged/submitted to filter_panel_refactor branch? You are trying to push 109 commits to main
![image](https://github.com/insightsengineering/teal.slice/assets/133694481/a4e900d8-0e2f-43bf-b8db-b11e179b35ad)

m7pr commented:
@kartikeyakirar is there any way you can extend tests/ folder for ChoicesFilterState and LogicalFilterState?
You can also pull filter_panel_refactor + pull your feature branch from pilter_panel_refactor to reflect today's merges on your feature branch. 

Can you share what R version you used to run tests and can you tell if you run tests for teal.gallery?

kartikeyakirar commented:
Hey @m7pr, I wanted to give you an update on the PR. I have made changes to the PR's title and pulled the latest changes from the 'filter_panel_refactor' branch on the 'main' branch. Currently, I am running all the unit tests again to ensure everything is working as expected. Just to let you know, I am using R version 4.1.3 (2022-03-10). I am updating the PR in light of the changes made in PR [#290.](https://github.com/insightsengineering/teal.slice/pull/290)

m7pr commented:
@kartikeyakirar I also see you edited NEWS.md - I think this is not needed, as Dawid will extend NEWS.md when we finally merge things to `main` and he also takes responsibility to bump a package version as well.

m7pr commented:
@kartikeyakirar you sure we should pass multiple = multiple parameter to init_filter_state.Date or init_filter_state.numeric? It only affects ChoicesFilterState and LogicalFilterState

Any reason it is default to FALSE for DatetimeFilterState and default to NULL for ChoicesFilterState ?

kartikeyakirar commented:
@m7pr  To ensure consistent parameters, an argument is passed to all classes that inherit from Filterstate. By default, multiple values are set to NULL, and if a null value is passed to ChoicesFilterState and LogicalFilterState, it is set to TRUE.

However, I appreciate your observation regarding DatetimeFilterState. It should default to NULL in that case. Thank you for pointing it out, and I am pushing the update

m7pr commented:
Yeah, I think it is the S3 Generic method that requires to have consistent parameters over various overloaded functions `init_filter_state`. It's not only consistency but rather a formal requirement https://cran.r-project.org/doc/manuals/R-exts.html#Generic-functions-and-methods so I guess that would broke tests if multiple wasn't passed further to methods that not necessarily use it. 
![image](https://github.com/insightsengineering/teal.slice/assets/133694481/5548876a-837c-4947-941d-79cbfd9eadbd)

m7pr commented:
@kartikeyakirar please check this comment where I provided a short video (created with zoom, over zoom meeting where I shared the screen and recorded the meeting) on how I run the app (with the app code) after all the changes were provided. Would you be able to do the same in here? Maybe you can reuse the code that I used to run the app (I picked the code from here https://github.com/insightsengineering/teal.slice/issues/264#issuecomment-1550514442)

chlebowa commented on R/FilterStateLogical.R:
```diff
@@ -188,11 +199,17 @@ LogicalFilterState <- R6::R6Class( # nolint
+        } else if (n_choices == 1 && !choices) {
           call("!", private$get_varname_prefixed(dataname))
+        } else {
+          call("%in%", varname, choices)
```
shouldn't this be `call("%in%", private$get_varname_prefixed(dataname), choices)`?

kartikeyakirar replied:
updated

chlebowa commented on R/FilterStateLogical.R:
```diff
@@ -248,13 +266,9 @@ LogicalFilterState <- R6::R6Class( # nolint
-      } else if (!isTRUE(private$get_keep_na()) && private$na_count > 0) {
-        TRUE
-      } else if (all(private$choices_counts > 0)) {
+      } else if (!setequal(private$get_selected(), private$choices)) {
```
```suggestion
      } else if (isFALSE(all.equal(private$get_selected(), private$choices))) {
```

chlebowa commented on R/FilterStateLogical.R:
```diff
@@ -248,13 +266,9 @@ LogicalFilterState <- R6::R6Class( # nolint
-      } else if (private$get_selected() == FALSE && private$choices_counts["FALSE"] == 0L) {
-        TRUE
-      } else if (private$get_selected() == TRUE && private$choices_counts["TRUE"] == 0L) {
+      } else if (!isTRUE(private$get_keep_na()) && private$na_count > 0) {
```
```suggestion
      } else if (isFALSE(private$get_keep_na()) && private$na_count > 0) {
```

chlebowa commented on R/FilterStateRange.R:
```diff
@@ -136,6 +136,8 @@ RangeFilterState <- R6::R6Class( # nolint
     #' @param choices (`atomic`, `NULL`)\cr
     #'   vector specifying allowed selection values
+    #' @param multiple (`logical(1)`)\cr
+    #'   flag specifying whether the `FilterState` is initiated multiple
```
```suggestion
    #'   flag specifying whether the `FilterState` more than one value can be selected;
    #'   only applicable to `FilterStateChoices` and `FilterStateLogical`
```

chlebowa commented on R/FilterStateRange.R:
```diff
@@ -169,6 +172,7 @@ RangeFilterState <- R6::R6Class( # nolint
                           ...) {
       checkmate::assert_numeric(x, all.missing = FALSE)
       checkmate::assert_numeric(choices, null.ok = TRUE)
+      checkmate::assert_flag(multiple, null.ok = TRUE)
```
```suggestion
```

chlebowa commented on R/teal_slice.R:
```diff
@@ -52,6 +52,8 @@
 #' @param choices optional vector specifying allowed choices;
 #'  possibly a subset of values in data; type and size depends on variable type
+#' @param multiple (`logical(1)`)\cr
+#'   flag specifying whether the `FilterState` is initiated multiple
```
```suggestion
    #'   flag specifying whether the `FilterState` more than one value can be selected;
    #'   only applicable to `FilterStateChoices` and `FilterStateLogical`
```

chlebowa commented on R/FilterState.R:
```diff
@@ -456,7 +463,7 @@ FilterState <- R6::R6Class( # nolint
         )
       )
-
+      private$set_na_rm(!value)
```
Why is this here now?

chlebowa reviewed (changes requested):
Some more tweaks necessary.
I will go through the test file later.

gogonzo commented on R/FilterStateChoices.R:
```diff
@@ -494,7 +520,7 @@ ChoicesFilterState <- R6::R6Class( # nolint
           } else {
             private$observers$selection <- observeEvent(
-              ignoreNULL = FALSE, # it's possible that nothing is selected
+              ignoreNULL = private$init_null, # init_null toggle for multiple var.
```
I think we should avoid no-selection when `multiple = FALSE`. Default behaviour of the filter-panel is as follows:
If `selected` is not specified then `selected` is set to `choices`, With `multiple = FALSE` we need to initialize filter with `selected = choices[1]`. `selected = NULL` when `multiple = FALSE` should be forbidden.

kartikeyakirar replied:
The variable private$init_null is set to TRUE when private$multiple is set to FALSE. This is because in the observeEvent for input$selection , the radio button, something has to be selected for input$selection, and it cannot be NULL. However, in the checkbox, it can be NULL. Hence, the variable private$init_null is used.

kartikeyakirar replied:
I have modified the code to ensure that if "multiple = FALSE" and no selection is made (selected = NULL), the first available choice will be automatically selected in this scenario.

kartikeyakirar commented on R/FilterStateChoices.R:
```diff
@@ -409,7 +435,7 @@ ChoicesFilterState <- R6::R6Class( # nolint
           choices = stats::setNames(private$choices, labels),
           selected = shiny::isolate(private$get_selected()),
-          multiple = TRUE,
+          multiple = private$multiple,
```
The "multiple" argument is passed to teal.widgets::optionalSelectInput, not to make_count_text().

chlebowa replied:
My mistake :+1:

kartikeyakirar commented on R/FilterStateChoices.R:
```diff
@@ -204,6 +210,11 @@ ChoicesFilterState <- R6::R6Class( # nolint
 
       private$set_choices_counts(unname(table(x_factor)))
 
+      if (private$multiple) {
+        private$init_null <- FALSE
+      } else {
+        private$init_null <- TRUE
+      }
```
The variable private$init_null is set to TRUE when private$multiple is set to FALSE. This is because in the observeEvent for input$selection , the radio button, something has to be selected for input$selection, and it cannot be NULL. However, in the checkbox, it can be NULL. Hence, the variable private$init_null is used.

gogonzo commented on R/FilterStateChoices.R:
```diff
@@ -494,7 +520,7 @@ ChoicesFilterState <- R6::R6Class( # nolint
           } else {
             private$observers$selection <- observeEvent(
-              ignoreNULL = FALSE, # it's possible that nothing is selected
+              ignoreNULL = private$init_null, # init_null toggle for multiple var.
```
@kartikeyakirar isn't this better?

```suggestion
              ignoreNULL = !private$multiple,
```

kartikeyakirar replied:
Indeed, @gogonzo, I had a similar idea, and it appears to be cleaner as well. However, I initially thought it might be confusing while reviewing the code, which led me to create the "init_null" variable. Nevertheless, I now believe that this approach is cleaner.

kartikeyakirar replied:
updated it and removed init_null

gogonzo commented on R/FilterStateLogical.R:
```diff
@@ -361,7 +393,7 @@ LogicalFilterState <- R6::R6Class( # nolint
 
           private$observers$selection <- observeEvent(
-            ignoreNULL = TRUE, # in radio button something has to be selected to input$selection can't be NULL
+            ignoreNULL = private$init_null,
```
```suggestion
            ignoreNULL = !private$multiple,
```

chlebowa commented on R/FilterStateChoices.R:
```diff
@@ -201,7 +201,12 @@ ChoicesFilterState <- R6::R6Class( # nolint
       private$set_choices(choices)
-      private$set_selected(selected)
+
+      if (is.null(selected) && isFALSE(multiple)) {
```
at this stage `multiple` cannot be `NULL`

I wonder, should this check be in the constructor or in `$set_choices`?

chlebowa replied:
My mistake, I misread as `is.null(multiple)`.

chlebowa replied:
Should it then not be like this?
```
      if (is.null(selected)) {
        if (multiple) {
          private$set_selected(choices)
        } else {
          private$set_selected(choices[1])
        }
      } else {
        private$set_selected(selected)
      }
```

kartikeyakirar replied:
I have updated code to fix https://github.com/insightsengineering/teal.slice/pull/296#discussion_r1210319410 and optimized this logic as well.

chlebowa commented on R/FilterStateChoices.R:
```diff
@@ -552,12 +557,12 @@ ChoicesFilterState <- R6::R6Class( # nolint
                 if(private$multiple) {
```
```suggestion
                if (private$multiple) {
```

chlebowa commented on R/FilterStateChoices.R:
```diff
@@ -530,6 +529,12 @@ ChoicesFilterState <- R6::R6Class( # nolint
                     private$varname,
                     private$dataname
                   ))
+                  if(is.null(input$selection) && isFALSE(private$multiple)) {
+                    selection_state <- private$get_selected()[1]
+                  } else {
+                    selection_state <- as.logical(input$selection)
+                  }
+
```
is `selection_state` used at all?

kartikeyakirar replied:
My mistake. It appears that the code in question was not utilized but was instead residual code from my testing.

chlebowa commented on R/FilterStateChoices.R:
```diff
@@ -552,12 +557,12 @@ ChoicesFilterState <- R6::R6Class( # nolint
                   updateRadioButtons(
                     inputId = "selection",
-                    selected =  private$selected()
+                    selected =  private$get_selected()[1]
```
Is the index necessary? `radioButtons` always returns one value, therefore so will `$get_selected`.

chlebowa replied:
Why would `$get_selected` return more than one value if `multiple = FALSE`?

kartikeyakirar commented:
#### The branches in use:

`filter_panel_refactor@main `branch for` teal `


#### Tested with below app code

```{R}
funny_module <- function(label = "Filter states", datanames = "all") {
  checkmate::assert_string(label)
  module(
    label = label,
    filters = datanames,
    ui = function(id, ...) {
      ns <- NS(id)
      div(
        h2("The following filter calls are generated:"),
        verbatimTextOutput(ns("filter_states")),
        verbatimTextOutput(ns("filter_calls")),
        actionButton(ns("reset"), "reset_to_default")
      )
    },
    server = function(input, output, session, data, filter_panel_api) {
      checkmate::assert_class(data, "tdata")
      observeEvent(input$reset, set_filter_state(filter_panel_api, default_filters))
      output$filter_states <- renderText({
        logger::log_trace("rendering text1")
        filter_panel_api |>
          get_filter_state() |>
          yaml::as.yaml()
      })
      output$filter_calls <- renderText({
        logger::log_trace("rendering text2")
        attr(data, "code")()
      })
    }
  )
}

options(teal.log_level = "TRACE", teal.show_js_log = TRUE)
options("teal.bs_theme" = bslib::bs_theme(version = 4))
# options(shiny.trace = TRUE)
pkgload::load_all("teal.slice")
pkgload::load_all("teal")
library(scda)

ADSL <- synthetic_cdisc_data("latest")$adsl
ADSL$logical1 <- FALSE
ADSL$logical <- sample(c(TRUE, FALSE), size = nrow(ADSL), replace = TRUE)
ADSL$logical[sample(1:nrow(ADSL), size = 10)] <- NA



default_filters <- filter_settings(
  filter_var(dataname = "ADSL", varname = "logical", selected = NULL, fixed = F,multiple = F),
  filter_var(dataname = "ADSL", varname = "logical1", selected = F, fixed = F,multiple = T),
  filter_var(dataname = "ADSL", varname = "SEX",     selected = c("F", "M"), multiple = F),
  filter_var(dataname = "ADSL", varname = "ETHNIC",    selected = NULL, multiple = T)
)

data <- cdisc_data(
  cdisc_dataset("ADSL", ADSL)
)

app <- teal::init(
  data = data,
  modules = list(funny_module(), funny_module("module2")),
  filter = default_filters
)

runApp(app)
```

#### for below R session

```{R}
> sessionInfo()
R version 4.1.3 (2022-03-10)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19042)

Matrix products: default

locale:
[1] LC_COLLATE=English_India.1252  LC_CTYPE=English_India.1252    LC_MONETARY=English_India.1252 LC_NUMERIC=C                  
[5] LC_TIME=English_India.1252    

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] teal.slice_0.3.0            MultiAssayExperiment_1.20.0 SummarizedExperiment_1.24.0 Biobase_2.54.0              GenomicRanges_1.46.1       
 [6] GenomeInfoDb_1.30.1         IRanges_2.28.0              S4Vectors_0.32.4            BiocGenerics_0.40.0         MatrixGenerics_1.6.0       
[11] matrixStats_0.61.0          BiocManager_1.30.16         teal_0.12.0.9026            teal.transform_0.2.0.9013   magrittr_2.0.3             
[16] devtools_2.4.3              usethis_2.1.6               teal.data_0.1.2.9013        scda_0.1.6.9000             testthat_3.1.8             
[21] shiny_1.7.4                

loaded via a namespace (and not attached):
 [1] colorspace_2.1-0         ellipsis_0.3.2           rsconnect_0.8.25         rprojroot_2.0.3          XVector_0.34.0          
 [6] fs_1.5.2                 rstudioapi_0.13          remotes_2.4.2            waldo_0.5.1              fansi_1.0.4             
[11] cachem_1.0.7             knitr_1.42               pkgload_1.2.4            jsonlite_1.8.4           compiler_4.1.3          
[16] backports_1.4.1          Matrix_1.5-4             fastmap_1.1.1            cli_3.6.1                later_1.3.0             
[21] htmltools_0.5.5          prettyunits_1.1.1        tools_4.1.3              gtable_0.3.3             glue_1.6.2              
[26] GenomeInfoDbData_1.2.7   dplyr_1.1.2              Rcpp_1.0.10              jquerylib_0.1.4          vctrs_0.6.1             
[31] teal.logger_0.1.1.9004   xfun_0.39                ps_1.6.0                 brio_1.1.3               mime_0.12               
[36] lifecycle_1.0.3          zlibbioc_1.40.0          scales_1.2.1             promises_1.2.0.1         yaml_2.3.5              
[41] memoise_2.0.1            ggplot2_3.4.2            teal.widgets_0.2.0.9025  sass_0.4.5               desc_1.4.1              
[46] checkmate_2.1.0          teal.reporter_0.1.1.9013 pkgbuild_1.3.1           rlang_1.1.0              pkgconfig_2.0.3         
[51] bitops_1.0-7             evaluate_0.21            lattice_0.20-45          fontawesome_0.5.1        purrr_1.0.1             
[56] processx_3.8.1           tidyselect_1.2.0         logger_0.2.2             R6_2.5.1                 generics_0.1.3          
[61] DelayedArray_0.20.0      pillar_1.9.0             withr_2.5.0              RCurl_1.98-1.6           tibble_3.2.1            
[66] crayon_1.5.2             shinyWidgets_0.7.6       utf8_1.2.3               rmarkdown_2.21           grid_4.1.3              
[71] callr_3.7.3              formatters_0.5.0         digest_0.6.31            xtable_1.8-4             httpuv_1.6.9            
[76] munsell_0.5.0            bslib_0.4.2              sessioninfo_1.2.2        shinyjs_2.1.0
```

gogonzo reviewed (commented):
It is still possible to set `selected` with multiple values even if `multiple = FALSE`. In UI everything is fine but calling `set_state` directly have a potential to break the app. There is a methods `validate_selection` and `remove_out_of_bound`. One of these should throw warning and limit selection only to one value.


```r
data_choices <- c(sample(letters[1:4], 100, replace = TRUE), NA)
attr(data_choices, "label") <- "lowercase letters"
fs <- ChoicesFilterState$new(
  x = data_choices,
  dataname = "data",
  varname = "variable",
  selected = c("a", "b"),
  keep_na = TRUE,
  multiple = FALSE
)
fs$set_state(filter_var(dataname = "data", varname = "variable", selected = c("a", "b")))
shiny::isolate(fs$get_state())

```

chlebowa commented on R/FilterStateChoices.R:
```diff
@@ -355,7 +352,21 @@ ChoicesFilterState <- R6::R6Class( # nolint
-      values[in_choices_mask]
+      values <- values[in_choices_mask]
+
+      if (is.null(values)) {
+        values <- if (private$multiple) private$choices else private$choices[1]
+      } else {
+        if (length(values) > 1 && !private$multiple) {
+          warning(paste(
+            "Values:", strtrim(paste(values, collapse = ", "), 360),
+            "are not a vector of length one.",
+            "The first value will be selected by default."
+          ))
+        }
+        values <- if (private$multiple) values else values[1]
+      }
```
1. Avoid using `paste` within condition functions.
2. Enumerating values is not necessary here.
```suggestion
          warning("\"values\" is not a vector of length one. The first value will be selected.")
```

I am not sure about the logic here. `if (is.null(value)) value <- private$choices` is done in `FilterState$set_selected` before calling this method, so there is no need to do it again. Something like 
'''
values <- 
  if (private$multiple) {
    values 
  } else {
    warning("warning message")
    values[1L]
  }
```
should be enough.

Also, it should be placed before checking for masked choices.

kartikeyakirar replied:
@chlebowa I removed mentioned logic i.e `if (is.null(value)) value <- private$choices`  before calling set_method.  and this has beed added to remove_out_of_bound_values() so there is no duplication same logic. 
https://github.com/insightsengineering/teal.slice/blob/279_multiple%40filter_panel_refactor%40main/R/FilterStateChoices.R#L204

kartikeyakirar replied:
A warning message, `"warning("\"values\" is not a vector of length one. The first value will be selected."),"` has been positioned after the masked choice. This placement is essential because if the user provides incorrect choices, such as in the following scenario:
`chars <- c("item1", "item2", "item3")
filter_state <- ChoicesFilterState$new(chars, dataname = "data", varname = "variable", multiple = FALSE)
    shiny::isolate(
      filter_state$set_state(filter_var(dataname = "data", varname = "variable", selected = c("item6", "item3")))
    )}`

In this case, "item6" will be set as the default value, even though it is not among the available choices. Consequently, it will be filtered out due to the masked choices, resulting in a null return value.

chlebowa commented on R/FilterStateLogical.R:
```diff
@@ -163,11 +168,7 @@ LogicalFilterState <- R6::R6Class( # nolint
       do.call(super$initialize, args)
 
       private$set_choices(choices)
-      if (is.null(selected)) {
-        private$set_selected(TRUE)
-      } else {
-        private$set_selected(selected)
-      }
```
I am not sure this version is better than the previous one. For logical values, the default selection is  `TURE`, not the alphabetically first one, which would be `FALSE`.

kartikeyakirar replied:
In the definition of the remove_out_bound function, I have set the default value to TRUE. You can find the code at this link: [GitHub link](https://github.com/insightsengineering/teal.slice/blob/279_multiple%40filter_panel_refactor%40main/R/FilterStateLogical.R#L250).

The logic of this function is similar to the remove_out_of_bound() function in filterStatechoice.R, with the only difference being that instead of using choice[1], I have set it to TRUE.

gogonzo commented on R/FilterStateLogical.R:
```diff
@@ -372,7 +410,12 @@ LogicalFilterState <- R6::R6Class( # nolint
                 )
               )
-              selection_state <- as.logical(input$selection)
+              if (is.null(input$selection) && isFALSE(private$multiple)) {
+                selection_state <- private$get_selected()
+              } else {
+                selection_state <- as.logical(input$selection)
+              }
```
It shouldn't be possible. When `private$multiple == TRUE` then we have `radioButton` so technically it's impossible to have nothing selected.

gogonzo commented on R/FilterStateChoices.R:
```diff
@@ -355,7 +352,21 @@ ChoicesFilterState <- R6::R6Class( # nolint
-      values[in_choices_mask]
+      values <- values[in_choices_mask]
+
+      if (is.null(values)) {
+        values <- if (private$multiple) private$choices else private$choices[1]
+      } else {
+        if (length(values) > 1 && !private$multiple) {
+          warning(paste(
+            "Values:", strtrim(paste(values, collapse = ", "), 360),
+            "are not a vector of length one.",
+            "The first value will be selected by default."
+          ))
+        }
```
```suggestion
      if (is.null(values)) {
        values <- if (private$multiple) private$choices else private$choices[1]
      } else if (length(values) > 1 && !private$multiple) {
         warning(
            "Values: ", strtrim(toString(values), 360),
            " are not a vector of length one.",
            " The first value will be selected by default."
          )
```

gogonzo commented on R/FilterStateChoices.R:
```diff
@@ -494,7 +516,7 @@ ChoicesFilterState <- R6::R6Class( # nolint
           } else {
             private$observers$selection <- observeEvent(
-              ignoreNULL = FALSE, # it's possible that nothing is selected
+              ignoreNULL = !private$mulitple, # toggle for multiple variable as radio button something has to be selected.
```
```suggestion
              ignoreNULL = !private$multiple, # toggle for multiple variable as radio button something has to be selected.
```

gogonzo commented on R/FilterStateChoices.R:
```diff
@@ -355,7 +352,21 @@ ChoicesFilterState <- R6::R6Class( # nolint
-      values[in_choices_mask]
+      values <- values[in_choices_mask]
+
+      if (is.null(values)) {
+        values <- if (private$multiple) private$choices else private$choices[1]
+      } else {
+        if (length(values) > 1 && !private$multiple) {
+          warning(paste(
+            "Values:", strtrim(paste(values, collapse = ", "), 360),
+            "are not a vector of length one.",
+            "The first value will be selected by default."
+          ))
+        }
+        values <- if (private$multiple) values else values[1]
+      }
```
1. is.null(values) will never be TRUE because of [this](https://github.com/insightsengineering/teal.slice/blob/e61bca49b7a3f12f64519a4c22838b743a5dd2a0/R/FilterState.R#L416)
2. This means that we just need to handle the situation when `multiple = FALSE` and `length(value) != 1`

I propose for this to have set_selected in FilterStateChoices and remove all the extra code you introduced here. `FilterStateChoices$set_selected` could look like:
```r
set_selected = function(value) {
  if (length(value) != 1) {
    warning("same warning as you wrote here")
    value <- private$choices[1]
  }
  super$set_selected(value)
}
```

kartikeyakirar replied:
This fixed. modified observeEvent for input$selection.

kartikeyakirar commented on R/FilterStateLogical.R:
```diff
@@ -248,13 +270,11 @@ LogicalFilterState <- R6::R6Class( # nolint
         TRUE
-      } else if (!isTRUE(private$get_keep_na()) && private$na_count > 0) {
-        TRUE
       } else if (all(private$choices_counts > 0)) {
         TRUE
-      } else if (private$get_selected() == FALSE && private$choices_counts["FALSE"] == 0L) {
+      } else if (isFALSE(all.equal(private$get_selected(), private$choices))) {
         TRUE
-      } else if (private$get_selected() == TRUE && private$choices_counts["TRUE"] == 0L) {
+      } else if (isFALSE(private$get_keep_na()) && private$na_count > 0) {
         TRUE
       } else {
         FALSE
```
I have restored the` !isTrue` 
and for second condition have added condition
`isFALSE(all.equal(private$get_selected(), private$choices)) &&
                 !anyNA(private$get_selected(), private$choices)`

kartikeyakirar commented on R/FilterStateLogical.R:
```diff
@@ -285,14 +305,25 @@ LogicalFilterState <- R6::R6Class( # nolint
+          inputId = ns("selection"),
+          label = NULL,
+          selected = shiny::isolate(as.character(private$get_selected())),
+          choiceNames = labels,
+          choiceValues = as.character(private$choices),
```
done

gogonzo commented on R/FilterStateChoices.R:
```diff
@@ -355,7 +352,13 @@ ChoicesFilterState <- R6::R6Class( # nolint
+      values <- values[in_choices_mask]
+
+      if (length(values) != 1 && !private$multiple) {
+        warning("\"values\" is not a vector of length one. The first value will be selected.")
```
Please fix this message to indicate which `dataname` and `varname` it refers to
```suggestion
        warning("\"values\" is not a vector of length one. The first value will be selected.")
```

gogonzo replied:
@kartikeyakirar do you mind to change the message to something more meaningful?

gogonzo commented on R/FilterStateChoices.R:
```diff
@@ -477,39 +491,34 @@ ChoicesFilterState <- R6::R6Class( # nolint
+              selection <- if (is.null(input$selection) && private$multiple){
+                character(0)
+              } else if (is.null(input$selection) && !private$multiple) {
+                showNotification("single selection is required, choice is ignored")
```
feel free to change this message to anything which will help user to understand what should be done 
```suggestion
                showNotification("Choice is ignored as this filter is restricted to single selection only.")
```

gogonzo commented on R/FilterStateChoices.R:
```diff
@@ -477,39 +491,34 @@ ChoicesFilterState <- R6::R6Class( # nolint
+
+              selection <- if (is.null(input$selection) && private$multiple){
+                character(0)
+              } else if (is.null(input$selection) && !private$multiple) {
```
We need to prevent from "select all"
```suggestion
              } else if (length(input$selection) != 1 && !private$multiple) {
```

gogonzo commented on R/FilterStateLogical.R:
```diff
@@ -213,10 +220,10 @@ LogicalFilterState <- R6::R6Class( # nolint
     },
     validate_selection = function(value) {
-      if (!(checkmate::test_logical(value, max.len = 1, any.missing = FALSE))) {
+      if (!(checkmate::test_logical(value, max.len = 2, any.missing = FALSE))) {
```
because
```r
checkmate::test_logical(c(TRUE, TRUE), max.len = 2, any.missing = FALSE)
#> TRUE
```


```suggestion
      if (!(checkmate::test_logical(value, any.missing = FALSE, unique = TRUE))) {
```

gogonzo commented on R/FilterStateLogical.R:
```diff
@@ -248,13 +262,12 @@ LogicalFilterState <- R6::R6Class( # nolint
       } else if (all(private$choices_counts > 0)) {
         TRUE
-      } else if (private$get_selected() == FALSE && private$choices_counts["FALSE"] == 0L) {
+      } else if (isFALSE(all.equal(private$get_selected(), private$choices)) &&
```
We don't want to flag as "filtered" when everything is selected in the different order
```r
all.equal(c(FALSE, TRUE), c(TRUE, FALSE)) 
#> [1] "2 element mismatches"

setequal(c(FALSE, TRUE), c(TRUE, FALSE)) 
#> [1] TRUE
```

```suggestion
      } else if (setequal(private$get_selected(), private$choices) &&
```

gogonzo replied:
We don't want to flag as "filtered" when everything is selected in the different order
```r
all.equal(c(FALSE, TRUE), c(TRUE, FALSE)) 
#> [1] "2 element mismatches"

setequal(c(FALSE, TRUE), c(TRUE, FALSE)) 
#> [1] TRUE
```

```suggestion
      } else if (setequal(private$get_selected(), private$choices) &&
```

gogonzo commented on R/utils.R:
```diff
@@ -125,3 +125,24 @@ include_js_files <- function(pattern) {
+#' # [1] 1 2 3
+#'
+#' make_c_call(1)
+#' # [1] 1
```
```suggestion
#' # [1] 1
#' @keywords internal
```

gogonzo commented on R/FilterStateChoices.R:
```diff
@@ -477,39 +491,34 @@ ChoicesFilterState <- R6::R6Class( # nolint
+                private$dataname
+              ))
+
+              selection <- if (is.null(input$selection) && private$multiple){
```
lintr
```suggestion
              selection <- if (is.null(input$selection) && private$multiple) {
```

gogonzo commented on tests/testthat/test-ChoicesFilterState.R:
```diff
@@ -305,6 +306,24 @@ testthat::test_that("set_state sets intersection of choices and passed values",
+    filter_state <- ChoicesFilterState$new(chars, dataname = "data", varname = "variable", multiple = FALSE)
+    shiny::isolate(
+      filter_state$set_state(filter_var(dataname = "data", varname = "variable", selected = c("item1", "item3")))
+    )}
+  )
```
lintr
```suggestion
    )
  })
```

gogonzo reviewed (commented):
Still possible to "Select all" in the `optionalSelectInput` with `multiple = FALSE`. See the [comment](https://github.com/insightsengineering/teal.slice/pull/296/files#r1213942535)

gogonzo reviewed (changes requested):
In a shiny session everything work as expected but API throws unnecessary warnings and returns wrong results:

On init default values (selected = NULL) throws warning and it shouldn't.
```r
filter_state <- teal.slice:::ChoicesFilterState$new(
  x = c(LETTERS, NA),
  varname = "x",
  dataname = "data",
  multiple = FALSE
)
print(filter_state) # selected: A (expected)
```

setting selected to single value throws warning and reactive value remains unchanged
```r
filter_var(dataname = "data", varname = "x", selected = "b") |> filter_state$set_state()
print(filter_state) # selected: A (unexpected)
```

In the code of `remove_out_of_bounds` we set `selected <- private$choices[1]` - I think we should set `selected <- shiny::isolate(private$selected[1])` in that case to be consistend with behaviour in server.
```r
filter_var(dataname = "data", varname = "x", selected = c("a", "b")) |> filter_state$set_state()
print(filter_state)
```

gogonzo commented on .github/workflows/check.yaml:
```diff
@@ -10,9 +10,11 @@ on:
       - ready_for_review
     branches:
       - main
+      - filter_panel_refactor@main
```
why are you changing these values? I think it should be first changed in filter_panel_refactor@main and merged in to your branch.

kartikeyakirar replied:
I did not change it manually.  at last commit  [abaeaf2](https://github.com/insightsengineering/teal.slice/pull/296/commits/abaeaf252ee8dd98ea8a11c51a608f17a92ae7b6) this got updated.

gogonzo commented on R/FilterStateChoices.R:
```diff
@@ -355,7 +358,18 @@ ChoicesFilterState <- R6::R6Class( # nolint
+
+      if (length(values) != 1 && !private$multiple) {
+        warning(sprintf(
+          "Values: %s are not a vector of length one.The first value will be selected by default.
```
```suggestion
          "Values: %s are not a vector of length one. The first value will be selected by default.
```

gogonzo commented on R/FilterStateChoices.R:
```diff
@@ -195,6 +201,12 @@ ChoicesFilterState <- R6::R6Class( # nolint
+        selected <- private$choices
+      } else if (is.null(selected)) {
+        selected <- private$choices[1]
+      }
```
There is potential problem on init:
```r
ChoicesFilterState$new(x = letters, dataname = "data", varname = "var", selected = c("b", "c"), multiple = FALSE)
```

```suggestion
      } else if (length(selected) > 1 && !multiple) {
        warning(
          "FilterStateChoices allows selected to be length=1 when multiple is FALSE. ",
          "Only first value is taken."
        )
        selected <- selected[1]
      }
```

Please include this test after change
```r
testthat::test_that("ChoicesFilterState constructor forces single selected when multiple is FALSE", {
  testthat::expect_warning(
    state <- ChoicesFilterState$new(
      x = letters, dataname = "data", varname = "var", selected = c("b", "c"), multiple = FALSE
    )
   )
   testthat::expect_identical(
     shiny::isolate(state$get_state()$selected),
     "b"
   )
})
```

gogonzo commented on R/FilterStateLogical.R:
```diff
@@ -163,11 +168,12 @@ LogicalFilterState <- R6::R6Class( # nolint
-        private$set_selected(selected)
+      if (is.null(selected) && multiple) {
+        selected <- private$choices
+      } else if (is.null(selected)) {
+        selected <- TRUE
```
This one gives an error
```r
LogicalFilterState$new(x = c(TRUE, FALSE), selected = c(TRUE, FALSE), multiple = FALSE)
```
```suggestion
      } else if (length(selected) != 1) {
        selected <- TRUE
```

Please include similar test to the one you've just added in test-ChoicesFilterState

gogonzo commented on R/FilterStateLogical.R:
```diff
@@ -148,11 +151,13 @@ LogicalFilterState <- R6::R6Class( # nolint
       checkmate::assert_class(x_reactive, "reactive")
       checkmate::assert_flag(selected, null.ok = TRUE)
```
```suggestion
      checkmate::assert_logical(selected, null.ok = TRUE)
```
